In [1]:
import pandas as pd #import pandas library to notebook

In [2]:
pd.set_option('display.max_columns', 200) # Sets the display columns in a dataframe to the number specified
pd.set_option('display.max_rows', 200) # sets the displays rows in a dataframe to the number specified

In [4]:
userInfo = pd.read_excel("User_Info.xlsx") #read_excel method in pandas is reading excel file from local to a dataframe

In [5]:
userInfo.head()

RESPONSE_ID  USER_ID  CUSTOMER_ID CRM_PRODUCT_TIER SALES_PRODUCT_TIER  \
0     14248134      803      1473270             free               free   
1     13502120      804      4641168             free                pro   
2     13954226      805      5654010             free                pro   
3     14315322      806      4802412             free                pro   
4     13042600      807      6746619             free                pro   

  MARKETING_PRODUCT_TIER PRODUCT_ADDONS SCREEN_SIZE  \
0                    pro          sites    1366x768   
1                    pro      reporting    1440x900   
2             enterprise  reporting|ads    1280x800   
3                    pro      reporting    1440x900   
4                    pro  reporting|ssl   1920x1080   

                   USER_ROLE LANGUAGE  
0  Especialista en marketing    es-mx  
1                        NaN    en-us  
2                Salesperson    en-us  
3                        NaN       en  
4     Spécialiste marketing    fr-fr

In [6]:
userInfo.columns #retreive the column headers available in the dataset

Index(['RESPONSE_ID', 'USER_ID', 'CUSTOMER_ID', 'CRM_PRODUCT_TIER',
       'SALES_PRODUCT_TIER', 'MARKETING_PRODUCT_TIER', 'PRODUCT_ADDONS',
       'SCREEN_SIZE', 'USER_ROLE', 'LANGUAGE'],
      dtype='object')

In [7]:
userInfo.dtypes # displays the datasets for all the columns in the dataset

RESPONSE_ID                int64
USER_ID                    int64
CUSTOMER_ID                int64
CRM_PRODUCT_TIER          object
SALES_PRODUCT_TIER        object
MARKETING_PRODUCT_TIER    object
PRODUCT_ADDONS            object
SCREEN_SIZE               object
USER_ROLE                 object
LANGUAGE                  object
dtype: object

In [8]:
userInfo.isnull() #Identify the missed values and this method will return a boolean at cell level based on value exists or not

RESPONSE_ID  USER_ID  CUSTOMER_ID  CRM_PRODUCT_TIER  SALES_PRODUCT_TIER  \
0           False    False        False             False               False   
1           False    False        False             False               False   
2           False    False        False             False               False   
3           False    False        False             False               False   
4           False    False        False             False               False   
...           ...      ...          ...               ...                 ...   
4783        False    False        False             False               False   
4784        False    False        False             False               False   
4785        False    False        False             False               False   
4786        False    False        False             False               False   
4787        False    False        False             False               False   

      MARKETING_PRODUCT_TIER  PRODUCT_ADDONS  SCREEN_SIZE  USER_ROLE  LANGUAGE  
0                      False           False        False      False     False  
1                      False           False        False       True     False  
2                      False           False        False      False     False  
3                      False           False        False       True     False  
4                      False           False        False      False     False  
...                      ...             ...          ...        ...       ...  
4783                   False            True        False      False     False  
4784                   False            True        False      False     False  
4785                   False            True        False       True     False  
4786                   False            True        False       True     False  
4787                   False            True        False      False     False  

[4788 rows x 10 columns]

In [9]:
user_miss_val = userInfo.isnull() # creating a new dataframe based on isnull method

In [10]:
user_miss_val

RESPONSE_ID  USER_ID  CUSTOMER_ID  CRM_PRODUCT_TIER  SALES_PRODUCT_TIER  \
0           False    False        False             False               False   
1           False    False        False             False               False   
2           False    False        False             False               False   
3           False    False        False             False               False   
4           False    False        False             False               False   
...           ...      ...          ...               ...                 ...   
4783        False    False        False             False               False   
4784        False    False        False             False               False   
4785        False    False        False             False               False   
4786        False    False        False             False               False   
4787        False    False        False             False               False   

      MARKETING_PRODUCT_TIER  PRODUCT_ADDONS  SCREEN_SIZE  USER_ROLE  LANGUAGE  
0                      False           False        False      False     False  
1                      False           False        False       True     False  
2                      False           False        False      False     False  
3                      False           False        False       True     False  
4                      False           False        False      False     False  
...                      ...             ...          ...        ...       ...  
4783                   False            True        False      False     False  
4784                   False            True        False      False     False  
4785                   False            True        False       True     False  
4786                   False            True        False       True     False  
4787                   False            True        False      False     False  

[4788 rows x 10 columns]

In [11]:
user_miss_val.columns.values.tolist()

['RESPONSE_ID',
 'USER_ID',
 'CUSTOMER_ID',
 'CRM_PRODUCT_TIER',
 'SALES_PRODUCT_TIER',
 'MARKETING_PRODUCT_TIER',
 'PRODUCT_ADDONS',
 'SCREEN_SIZE',
 'USER_ROLE',
 'LANGUAGE']

In [12]:
user_miss_val["USER_ROLE"].value_counts() # returns the count for distinct values existing in the columns

False    2930
True     1858
Name: USER_ROLE, dtype: int64

**Count the missing values in each column. "True" respresents the missing values and "False" represnts the values exists in Dataset**

In [14]:
for col_name in user_miss_val.columns.values.tolist(): # taking each column name into the existing variable
    print("Column Name", col_name)                     # within the loop we will be print each column name
    print(user_miss_val[col_name].value_counts())      # each column name to the method to get distinct value counts
    print('----------------------------------------------') # Line seperator

Column Name RESPONSE_ID
False    4788
Name: RESPONSE_ID, dtype: int64
----------------------------------------------
Column Name USER_ID
False    4788
Name: USER_ID, dtype: int64
----------------------------------------------
Column Name CUSTOMER_ID
False    4788
Name: CUSTOMER_ID, dtype: int64
----------------------------------------------
Column Name CRM_PRODUCT_TIER
False    4788
Name: CRM_PRODUCT_TIER, dtype: int64
----------------------------------------------
Column Name SALES_PRODUCT_TIER
False    4788
Name: SALES_PRODUCT_TIER, dtype: int64
----------------------------------------------
Column Name MARKETING_PRODUCT_TIER
False    4788
Name: MARKETING_PRODUCT_TIER, dtype: int64
----------------------------------------------
Column Name PRODUCT_ADDONS
True     2892
False    1896
Name: PRODUCT_ADDONS, dtype: int64
----------------------------------------------
Column Name SCREEN_SIZE
False    4788
Name: SCREEN_SIZE, dtype: int64
----------------------------------------------
Column

**User Info Data Set : Missing data findings** 
Based on the summary above, out of 4788 rows of data, and 2 columns containing missing data:

1. "PRODUCT_ADDONS": 2892 missing data
2. "USER_ROLE": 1858 missing data

**Data Analysis**

In [16]:
userInfo.columns

Index(['RESPONSE_ID', 'USER_ID', 'CUSTOMER_ID', 'CRM_PRODUCT_TIER',
       'SALES_PRODUCT_TIER', 'MARKETING_PRODUCT_TIER', 'PRODUCT_ADDONS',
       'SCREEN_SIZE', 'USER_ROLE', 'LANGUAGE'],
      dtype='object')

In [17]:
userInfo[['USER_ID','PRODUCT_ADDONS','SCREEN_SIZE','USER_ROLE']].describe(include='all')
#descriptive sets in the given dataset based on the listed coulmns 
#using describe method from Pandas Library to achieve that

USER_ID PRODUCT_ADDONS SCREEN_SIZE               USER_ROLE
count   4788.000000           1896        4788                    2930
unique          NaN             12         130                      41
top             NaN      reporting   1920x1080  Executive / Management
freq            NaN            867        1203                     844
mean    2394.500000            NaN         NaN                     NaN
std     1382.320874            NaN         NaN                     NaN
min        1.000000            NaN         NaN                     NaN
25%     1197.750000            NaN         NaN                     NaN
50%     2394.500000            NaN         NaN                     NaN
75%     3591.250000            NaN         NaN                     NaN
max     4788.000000            NaN         NaN                     NaN

In [18]:
userInfo[['USER_ID','PRODUCT_ADDONS','SCREEN_SIZE','USER_ROLE']]

USER_ID PRODUCT_ADDONS SCREEN_SIZE                       USER_ROLE
0         803          sites    1366x768       Especialista en marketing
1         804      reporting    1440x900                             NaN
2         805  reporting|ads    1280x800                     Salesperson
3         806      reporting    1440x900                             NaN
4         807  reporting|ssl   1920x1080          Spécialiste marketing
...       ...            ...         ...                             ...
4783     4784            NaN   2560x1440          Executive / Management
4784     4785            NaN    1366x768  Partner / Agencia / Freelancer
4785     4786            NaN    1280x720                             NaN
4786     4787            NaN    1366x768                             NaN
4787     4788            NaN    1600x900                           Other

[4788 rows x 4 columns]

In [20]:
a = userInfo.sort_values(by=['CUSTOMER_ID'])
a

RESPONSE_ID  USER_ID  CUSTOMER_ID CRM_PRODUCT_TIER SALES_PRODUCT_TIER  \
507      13172764       13         2973             none               none   
319      13174384     1122        30462             free                pro   
3754     13017628     3453        31341             free                pro   
1582     13857259     2769        31341             free                pro   
1759     12926630     1463        32100             free               free   
...           ...      ...          ...              ...                ...   
1916     14257156     1620     12939510             free               free   
4597     14200413     4598     12950106             free               free   
729      13894482      235     12952059             free               free   
3605     13948750     3304     12965613             free               free   
406      14306118     1209     13012215             free            starter   

     MARKETING_PRODUCT_TIER PRODUCT_ADDONS SCREEN_SIZE  \
507                    none            NaN   1920x1080   
319              enterprise  reporting|ads   2313x1446   
3754                    pro            NaN    1366x768   
1582                    pro            NaN   1920x1080   
1759             enterprise  reporting|ads   1920x1080   
...                     ...            ...         ...   
1916                   free            NaN   1920x1200   
4597                   free            NaN   1920x1080   
729                    free            NaN   1920x1080   
3605                   free            NaN   1500x1000   
406                    free            NaN   1920x1080   

                   USER_ROLE LANGUAGE  
507                      NaN       en  
319                      NaN       en  
3754                     NaN    en-us  
1582                     NaN    en-us  
1759                Marketer       en  
...                      ...      ...  
1916  Executive / Management       en  
4597  Executive / Management       en  
729                 Marketer       en  
3605                     NaN       en  
406   Executive / Management    en-gb  

[4788 rows x 10 columns]

In [25]:
a = userInfo.sort_values(by=['USER_ID'],ascending=False)
a

RESPONSE_ID  USER_ID  CUSTOMER_ID CRM_PRODUCT_TIER SALES_PRODUCT_TIER  \
4787     14313509     4788      8014902             free               free   
4786     14355517     4787     12808452             free            starter   
4785     14346754     4786     12687168             free               free   
4784     14202834     4785     12189894             free               free   
4783     14350482     4784      7157487             free                pro   
...           ...      ...          ...              ...                ...   
499      12940047        5       205140             free                pro   
498      13235515        4       567072             free               free   
497      12880156        3       109137             free                pro   
496      13291804        2      1143396             free                pro   
495      12883314        1        62175             none               none   

     MARKETING_PRODUCT_TIER PRODUCT_ADDONS SCREEN_SIZE  \
4787                   free            NaN    1600x900   
4786                   free            NaN    1366x768   
4785                   free            NaN    1280x720   
4784                   free            NaN    1366x768   
4783                starter            NaN   2560x1440   
...                     ...            ...         ...   
499              enterprise            NaN   1920x1080   
498              enterprise            ads   1920x1080   
497              enterprise            ads    1280x800   
496                     pro            NaN   1920x1080   
495                    none            NaN    1600x900   

                           USER_ROLE LANGUAGE  
4787                           Other    en-us  
4786                             NaN       en  
4785                             NaN       en  
4784  Partner / Agencia / Freelancer    es-mx  
4783          Executive / Management       en  
...                              ...      ...  
499    Partner / Agency / Freelancer       en  
498    Partner / Agency / Freelancer       en  
497           Executive / Management       nl  
496                              NaN    en-us  
495                              NaN       en  

[4788 rows x 10 columns]

In [26]:
a.to_excel('writeFile.xlsx')

In [27]:
readFile = pd.read_excel('writeFile.xlsx')

In [28]:
readFile.head()

Unnamed: 0  RESPONSE_ID  USER_ID  CUSTOMER_ID CRM_PRODUCT_TIER  \
0        4787     14313509     4788      8014902             free   
1        4786     14355517     4787     12808452             free   
2        4785     14346754     4786     12687168             free   
3        4784     14202834     4785     12189894             free   
4        4783     14350482     4784      7157487             free   

  SALES_PRODUCT_TIER MARKETING_PRODUCT_TIER PRODUCT_ADDONS SCREEN_SIZE  \
0               free                   free            NaN    1600x900   
1            starter                   free            NaN    1366x768   
2               free                   free            NaN    1280x720   
3               free                   free            NaN    1366x768   
4                pro                starter            NaN   2560x1440   

                        USER_ROLE LANGUAGE  
0                           Other    en-us  
1                             NaN       en  
2                             NaN       en  
3  Partner / Agencia / Freelancer    es-mx  
4          Executive / Management       en